service/history/execution/context.go

```diff
@@ -265,11 +270,27 @@ func NewContext(
 }
 
 func (c *contextImpl) Lock(ctx context.Context) error {
-	return c.mutex.Lock(ctx)
+	err := c.mutex.Lock(ctx)
+	if err != nil {
+		return err
+	}
+	c.lockTime.Store(time.Now().UnixNano())
+	return nil
 }
 
 func (c *contextImpl) Unlock() {
+	lockTime := c.lockTime.Swap(0)
 	c.mutex.Unlock()
+
+	if lockTime == 0 { // skip logging if the lock is never acquired
+		return
+	}
+	elapsed := time.Since(time.Unix(0, lockTime))
+	c.metricsClient.RecordTimer(metrics.WorkflowContextScope, metrics.WorkflowContextLockLatency, elapsed)
+	if elapsed > c.maxLockDuration {
+		c.maxLockDuration = elapsed
```

Comment:
read and mutated concurrently outside the lock

Reply:
missed it. better to do things when lock is not released

Reply:
yea.  `defer`'s fine I think.  the log should be pretty cheap since it's not serializing anything big.